File: internal/connector/selector.go
package connector

import (
	"encoding/json"
	"strconv"
	"strings"
)

// Select evaluates a small path expression against decoded JSON and returns
// every string it matched.
//
// Supported: dot notation, [*] to expand an array, [n] to index one, and
// [key] / [!key] to keep only elements where a field is truthy / falsy. That is
// deliberately less than JSONPath — it covers real response shapes without
// importing an engine whose edge cases we would then own.
//
//	$.content.parts[*].text
//	$.content.parts[!thought].text     reasoning models: skip the thinking
//	$.choices[0].delta.content
//	$.output
//
// The [!key] form exists because reasoning models interleave their thinking
// with their answer in the same array, flagged by a field. Without it a
// connector concatenates the model's scratchpad into the reply.
func Select(data any, path string) []string {
	steps := parsePath(path)
	if len(steps) == 0 {
		return stringify(data)
	}
	cur := []any{data}
	for _, st := range steps {
		var next []any
		for _, node := range cur {
			next = append(next, st.apply(node)...)
		}
		if len(next) == 0 {
			return nil
		}
		cur = next
	}
	var out []string
	for _, n := range cur {
		out = append(out, stringify(n)...)
	}
	return out
}

type step struct {
	key     string
	index   int  // used when isIdx
	isAll   bool // [*]
	isIdx   bool // [n]
	hasKey  bool
	predKey string // [field] or [!field]
	predNot bool   // true for [!field]
}

func (s step) apply(node any) []any {
	if s.hasKey {
		m, ok := node.(map[string]any)
		if !ok {
			return nil
		}
		v, ok := m[s.key]
		if !ok {
			return nil
		}
		node = v
	}
	switch {
	case s.predKey != "":
		arr, ok := node.([]any)
		if !ok {
			return nil
		}
		var out []any
		for _, e := range arr {
			if truthyField(e, s.predKey) != s.predNot {
				out = append(out, e)
			}
		}
		return out
	case s.isAll:
		arr, ok := node.([]any)
		if !ok {
			return nil
		}
		return arr
	case s.isIdx:
		arr, ok := node.([]any)
		if !ok || s.index < 0 || s.index >= len(arr) {
			return nil
		}
		return []any{arr[s.index]}
	default:
		return []any{node}
	}
}

// truthyField reports whether an element carries a truthy value at key. A
// missing field counts as false, so [!thought] keeps plain elements.
func truthyField(node any, key string) bool {
	m, ok := node.(map[string]any)
	if !ok {
		return false
	}
	switch v := m[key].(type) {
	case nil:
		return false
	case bool:
		return v
	case string:
		return v != "" && v != "false"
	case float64:
		return v != 0
	default:
		return true
	}
}

func parsePath(path string) []step {
	p := strings.TrimSpace(path)
	p = strings.TrimPrefix(p, "$")
	p = strings.TrimPrefix(p, ".")
	if p == "" {
		return nil
	}
	var steps []step
	for _, seg := range strings.Split(p, ".") {
		if seg == "" {
			continue
		}
		name := seg
		var brackets []string
		for {
			i := strings.Index(name, "[")
			if i < 0 {
				break
			}
			j := strings.Index(name[i:], "]")
			if j < 0 {
				break
			}
			j += i
			brackets = append(brackets, name[i+1:j])
			name = name[:i] + name[j+1:]
		}
		st := step{}
		if name != "" {
			st.key, st.hasKey = name, true
		}
		if len(brackets) == 0 {
			steps = append(steps, st)
			continue
		}
		// First bracket attaches to this key; further ones become bare steps.
		for bi, b := range brackets {
			s := st
			if bi > 0 {
				s = step{}
			}
			switch {
			case b == "*":
				s.isAll = true
			case strings.HasPrefix(b, "!"):
				s.predKey, s.predNot = strings.TrimPrefix(b, "!"), true
			default:
				if n, err := strconv.Atoi(b); err == nil {
					s.isIdx, s.index = true, n
				} else if b != "" {
					s.predKey = b
				}
			}
			steps = append(steps, s)
		}
	}
	return steps
}

func stringify(v any) []string {
	switch t := v.(type) {
	case nil:
		return nil
	case string:
		if t == "" {
			return nil
		}
		return []string{t}
	case float64:
		return []string{strconv.FormatFloat(t, 'f', -1, 64)}
	case bool:
		return []string{strconv.FormatBool(t)}
	case []any:
		var out []string
		for _, e := range t {
			out = append(out, stringify(e)...)
		}
		return out
	default:
		b, err := json.Marshal(v)
		if err != nil {
			return nil
		}
		return []string{string(b)}
	}
}

// SelectFirst tries each path in order and returns the first non-empty match.
// Returning the first match rather than concatenating all of them keeps a
// fallback selector from duplicating text the primary one already produced.
func SelectFirst(data any, paths []string) []string {
	for _, p := range paths {
		if got := Select(data, p); len(got) > 0 {
			return got
		}
	}
	return nil
}

// Truthy reports whether a path resolves to something meaning "yes", used for
// the optional done signal.
func Truthy(data any, path string) bool {
	if path == "" {
		return false
	}
	for _, s := range Select(data, path) {
		switch strings.ToLower(s) {
		case "", "false", "0", "null":
		default:
			return true
		}
	}
	return false
}

// Matches evaluates a condition against a payload. Three forms:
//
//	$.partial                        truthy
//	$.type == TEXT_MESSAGE_CONTENT   equal to
//	$.type != REASONING_CONTENT      not equal to
//
// Equality exists because protocols with typed events — AG-UI is the common
// case — distinguish an answer from reasoning by the value of a field rather
// than by its presence, and both carry the same payload shape otherwise.
func Matches(data any, cond string) bool {
	cond = strings.TrimSpace(cond)
	if cond == "" {
		return true
	}
	path, op, want := splitCondition(cond)
	if op == "" {
		return Truthy(data, path)
	}
	found := false
	for _, s := range Select(data, path) {
		if s == want {
			found = true
			break
		}
	}
	if op == "!=" {
		return !found
	}
	return found
}

func splitCondition(cond string) (path, op, want string) {
	for _, o := range []string{"==", "!="} {
		if i := strings.Index(cond, o); i >= 0 {
			path = strings.TrimSpace(cond[:i])
			want = strings.TrimSpace(cond[i+len(o):])
			want = strings.Trim(want, `"'`)
			return path, o, want
		}
	}
	return cond, "", ""
}

File: src/connector/selector.rs
use serde_json::Value;

#[derive(Clone, Debug, Default)]
struct Step {
    key: Option<String>,
    bracket: Bracket,
}

#[derive(Clone, Debug, Default)]
enum Bracket {
    #[default]
    None,
    All,
    Index(usize),
    Predicate {
        key: String,
        negated: bool,
    },
}

pub fn select(data: &Value, path: &str) -> Vec<String> {
    let steps = parse_path(path);
    if steps.is_empty() {
        return stringify(data);
    }
    let mut current = vec![data];
    for step in &steps {
        let mut next = Vec::new();
        for node in current {
            next.extend(apply(node, step));
        }
        if next.is_empty() {
            return Vec::new();
        }
        current = next;
    }
    current.into_iter().flat_map(stringify).collect()
}

pub fn select_first(data: &Value, paths: &[String]) -> Vec<String> {
    for path in paths {
        let selected = select(data, path);
        if !selected.is_empty() {
            return selected;
        }
    }
    Vec::new()
}

pub fn truthy(data: &Value, path: &str) -> bool {
    !path.is_empty()
        && select(data, path).into_iter().any(|value| {
            !matches!(
                value.to_ascii_lowercase().as_str(),
                "" | "false" | "0" | "null"
            )
        })
}

pub fn matches(data: &Value, condition: &str) -> bool {
    let condition = condition.trim();
    if condition.is_empty() {
        return true;
    }
    for operator in ["==", "!="] {
        if let Some(index) = condition.find(operator) {
            let path = condition[..index].trim();
            let wanted = condition[index + operator.len()..]
                .trim()
                .trim_matches(['\'', '"']);
            let found = select(data, path).iter().any(|value| value == wanted);
            return if operator == "!=" { !found } else { found };
        }
    }
    truthy(data, condition)
}

fn apply<'a>(node: &'a Value, step: &Step) -> Vec<&'a Value> {
    let node = match &step.key {
        Some(key) => match node.get(key) {
            Some(value) => value,
            None => return Vec::new(),
        },
        None => node,
    };
    match &step.bracket {
        Bracket::None => vec![node],
        Bracket::All => node
            .as_array()
            .map_or_else(Vec::new, |items| items.iter().collect()),
        Bracket::Index(index) => node.get(*index).into_iter().collect(),
        Bracket::Predicate { key, negated } => node.as_array().map_or_else(Vec::new, |items| {
            items
                .iter()
                .filter(|item| field_truthy(item, key) != *negated)
                .collect()
        }),
    }
}

fn field_truthy(node: &Value, key: &str) -> bool {
    node.get(key).is_some_and(value_truthy)
}

fn value_truthy(value: &Value) -> bool {
    match value {
        Value::Null => false,
        Value::Bool(value) => *value,
        Value::String(value) => !value.is_empty() && value != "false",
        Value::Number(value) => value.as_f64() != Some(0.0),
        _ => true,
    }
}

fn parse_path(path: &str) -> Vec<Step> {
    let path = path.trim().strip_prefix('$').unwrap_or(path.trim());
    let path = path.strip_prefix('.').unwrap_or(path);
    if path.is_empty() {
        return Vec::new();
    }
    let mut output = Vec::new();
    for segment in path.split('.').filter(|segment| !segment.is_empty()) {
        let mut name = segment.to_owned();
        let mut brackets = Vec::new();
        while let Some(start) = name.find('[') {
            let Some(relative_end) = name[start..].find(']') else {
                break;
            };
            let end = start + relative_end;
            brackets.push(name[start + 1..end].to_owned());
            name.replace_range(start..=end, "");
        }
        if brackets.is_empty() {
            output.push(Step {
                key: (!name.is_empty()).then_some(name),
                bracket: Bracket::None,
            });
            continue;
        }
        for (index, bracket) in brackets.into_iter().enumerate() {
            let key = (index == 0 && !name.is_empty()).then(|| name.clone());
            let bracket = match bracket.as_str() {
                "*" => Bracket::All,
                value if value.starts_with('!') => Bracket::Predicate {
                    key: value[1..].to_owned(),
                    negated: true,
                },
                value => match value.parse::<usize>() {
                    Ok(index) => Bracket::Index(index),
                    Err(_) => Bracket::Predicate {
                        key: value.to_owned(),
                        negated: false,
                    },
                },
            };
            output.push(Step { key, bracket });
        }
    }
    output
}

fn stringify(value: &Value) -> Vec<String> {
    match value {
        Value::Null => Vec::new(),
        Value::String(value) if value.is_empty() => Vec::new(),
        Value::String(value) => vec![value.clone()],
        Value::Number(value) => vec![value.to_string()],
        Value::Bool(value) => vec![value.to_string()],
        Value::Array(values) => values.iter().flat_map(stringify).collect(),
        value => serde_json::to_string(value).into_iter().collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn selector_parity() {
        let document = json!({
            "content": {"parts": [
                {"text": "thinking", "thought": true},
                {"text": "answer"}
            ]},
            "choices": [{"delta": {"content": "hi"}}],
            "n": 42
        });
        assert_eq!(
            select(&document, "$.content.parts[*].text"),
            ["thinking", "answer"]
        );
        assert_eq!(
            select(&document, "$.content.parts[!thought].text"),
            ["answer"]
        );
        assert_eq!(
            select(&document, "$.content.parts[thought].text"),
            ["thinking"]
        );
        assert_eq!(select(&document, "$.choices[0].delta.content"), ["hi"]);
        assert_eq!(select(&document, "$.n"), ["42"]);
        assert!(select(&document, "$.missing").is_empty());
    }

    #[test]
    fn condition_parity() {
        let text = json!({"type": "TEXT_MESSAGE_CONTENT", "partial": true});
        assert!(matches(&text, "$.type == TEXT_MESSAGE_CONTENT"));
        assert!(matches(&text, "$.type != REASONING_MESSAGE_CONTENT"));
        assert!(matches(&text, "$.partial"));
        assert!(!matches(&text, "$.missing"));
    }
}
